### DISCUSSION TITLE
How to use mark_end?

### DISCUSSION THREAD

**Discussion by Himujjal** [tree-sitter/tree-sitter#910] [Q&A] [answered]
I have a problem where I need the expression:
```ejs
{#await promise then value}
```
to be parsed as `(if_block (raw_text_await) (raw_text_expr))`. If `then` is not there, then it will be (if_block (raw_text_await) (raw_text_expr)) 

Here, I need `raw_text_await` to contain `"promise"` as text. So, I need to lookahead to get the then value. Here is my implementation in scanner.c.
```c
// ...
lexer->mark_end(lexer);
lexer->advance(lexer, false);
if (scan_then(lexer)) {
  lexer->mark_end(lexer);
  lexer->result_symbol = RAW_TEXT_AWAIT;
}
// ...
```
Its giving me results but its also including the `then` word in `raw_text_await`.

**Comment by ahlinc** [comment, marked as answer]
You don't need the second `lexer->mark_end(lexer);` call.
Try to do the next:
1. Consume the "promise" token.
2. Mark the "promise" end by calling `lexer->mark_end(lexer);`
3. Lookahead what you need and if ok just make `lexer->result_symbol = ...; return true;`
  3.1. If not ok then `return false;` without any other actions.
